depth_to_pointcloud/src/depth_to_pointcloud_node.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <image_geometry/pinhole_camera_model.h>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/image_encodings.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <sensor_msgs/point_cloud2_iterator.hpp>
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+static rclcpp::publisher::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr g_pub_point_cloud;
+
+static sensor_msgs::msg::CameraInfo::SharedPtr g_cam_info;
+
+// TODO(clalancette): The DepthTraits structure and the convert<>() function are
+// both copied from
+// https://github.com/ros-perception/image_pipeline/tree/indigo/depth_image_proc .
+// However, it is not clear that it is worthwhile to port depth_image_proc from
+// ROS1, since it is heavily tied to ROS1 nodelets.  For now, we leave the
+// copied code in here until we figure out what we want to do.
+
+// Encapsulate differences between processing float and uint16_t depths
+template<typename T>
+struct DepthTraits {};
+
+template<>
+struct DepthTraits<uint16_t>
+{
+  static inline bool valid(uint16_t depth) {return depth != 0; }
+  static inline float toMeters(uint16_t depth) {return depth * 0.001f; }  // originally mm
+  static inline uint16_t fromMeters(float depth) {return (depth * 1000.0f) + 0.5f; }
+  static inline void initializeBuffer(std::vector<uint8_t> &) {}  // Do nothing
+};
+
+template<>
+struct DepthTraits<float>
+{
+  static inline bool valid(float depth) {return std::isfinite(depth); }
+  static inline float toMeters(float depth) {return depth; }
+  static inline float fromMeters(float depth) {return depth; }
+
+  static inline void initializeBuffer(std::vector<uint8_t> & buffer)
+  {
+    float * start = reinterpret_cast<float *>(&buffer[0]);
+    float * end = reinterpret_cast<float *>(&buffer[0] + buffer.size());
+    std::fill(start, end, std::numeric_limits<float>::quiet_NaN());
+  }
+};
```

Comment:
Good point.  I think what I'll do here is split this out into a separate file with a BSD license, like the original.